vortex-serde/src/layouts/read/cache.rs

```diff
@@ -26,23 +32,117 @@ impl LayoutMessageCache {
     }
 }
 
+#[derive(Debug)]
+enum LazyDTypeState {
+    Value(DType),
+    Serialized(Bytes, OnceCell<DType>, Projection),
+}
+
+#[derive(Debug)]
+pub struct LazyDeserializedDType {
+    inner: LazyDTypeState,
+}
+
+impl LazyDeserializedDType {
+    pub fn from_bytes(dtype_bytes: Bytes, projection: Projection) -> Self {
+        Self {
+            inner: LazyDTypeState::Serialized(dtype_bytes, OnceCell::new(), projection),
+        }
+    }
+
+    pub fn from_dtype(dtype: DType) -> Self {
+        Self {
+            inner: LazyDTypeState::Value(dtype),
+        }
+    }
+
+    /// Restrict the underlying dtype to selected fields
+    pub fn project(&self, projection: &[Field]) -> VortexResult<Arc<Self>> {
+        match &self.inner {
+            LazyDTypeState::Value(d) => {
+                let DType::Struct(s, n) = d else {
+                    vortex_bail!("Not a struct dtype")
+                };
+                Ok(Arc::new(LazyDeserializedDType::from_dtype(DType::Struct(
+                    s.project(projection)?,
+                    *n,
+                ))))
+            }
+            // TODO(robert): Respect existing projection list, only really an issue for nested structs
+            LazyDTypeState::Serialized(b, ..) => Ok(Arc::new(LazyDeserializedDType::from_bytes(
+                b.clone(),
+                Projection::Flat(projection.to_owned()),
+            ))),
+        }
+    }
+
+    /// Get vortex dtype out of serialized bytes
+    pub fn value(&self) -> VortexResult<&DType> {
+        match &self.inner {
+            LazyDTypeState::Value(d) => Ok(d),
+            LazyDTypeState::Serialized(b, c, p) => c.get_or_try_init(|| {
```

Comment:
this is too many single character variable names, really hard to read without an IDE